solutions/platinum/ioi-05-mountain.mdx

```diff
@@ -7,7 +7,121 @@ author: Justin Ji
 
 [Official Analysis](https://dmoj.ca/problem/ioi05p3/editorial)
 
-## Explanation
+<Spoiler title = "Sparse Segment Tree Solution">
+
+This problem can be solved with a standard sparse segment tree. We keep track
+of the prefix max and range sum in each node, and walk on the segment tree to
+answer queries. The implementation below uses pointers, but it is possible to
+write the solution without using them.
+
+**Time Complexity:** $\mathcal{O}(Q\log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+class SparseSegtree {
+  private:
+	const int n;
+	struct Node {
+		// vals = {prefix max, range sum}
+		array<int, 2> vals = {0, 0};
+		int lazy_set = INT32_MAX;
+		Node *left = nullptr;
+		Node *right = nullptr;
+	};
+	Node *root = new Node;
+
+	inline array<int, 2> comb(const array<int, 2> &a, const array<int, 2> &b) {
+		return {max(a[0], a[1] + b[0]), a[1] + b[1]};
+	}
+
+	void apply(Node *cur, int len, int val) {
+		(cur->vals)[1] = len * val;
+		(cur->vals)[0] = max(0, len * val);
+		(cur->lazy_set) = val;
+	}
+
+	void push_down(Node *cur, int l, int r) {
+		if ((cur->left) == nullptr) { (cur->left) = new Node; }
+		if ((cur->right) == nullptr) { (cur->right) = new Node; }
+		if (l != r && (cur->lazy_set) != INT32_MAX) {
+			int m = (l + r) / 2;
+			apply(cur->left, m - l + 1, cur->lazy_set);
+			apply(cur->right, r - m, cur->lazy_set);
+		}
+		(cur->lazy_set) = INT32_MAX;
+	}
+
+	void range_set(Node *cur, int l, int r, int ql, int qr, int val) {
+		if (qr < l || ql > r) { return; }
+		if (ql <= l && r <= qr) {
+			apply(cur, r - l + 1, val);
+		} else {
+			push_down(cur, l, r);
+			int m = (l + r) / 2;
+			range_set(cur->left, l, m, ql, qr, val);
+			range_set(cur->right, m + 1, r, ql, qr, val);
+			(cur->vals) = comb((cur->left)->vals, (cur->right)->vals);
+		}
+	}
+
+	int walk_right(Node *cur, int l, int r, const array<int, 2> &pf, int mx) {
+		if (l == r) {
+			return l;
+		} else {
+			push_down(cur, l, r);
+			int m = (l + r) / 2;
+			array<int, 2> x = comb(pf, (cur->left)->vals);
+			return x[0] > mx ? walk_right(cur->left, l, m, pf, mx)
+			                 : walk_right(cur->right, m + 1, r, x, mx);
+		}
+	}
+
+  public:
+	SparseSegtree(int n) : n(n) {}
+
+	inline void range_set(int ql, int qr, int val) {
+		range_set(root, 0, n - 1, ql, qr, val);
+	}
+
+	inline int walk_right(int mx) {
```

Comment:
ok this defo doesn't matter but idt `inline` does anything here
it is just a hint to the compiler so

Reply:
fair enough, i can remove it later

Reply:
ok removed it, vscode said that the compiler was inlining it anyways lmao

Reply:
yeah the compiler just does whatever the f it wants